artichoke-backend/src/extn/core/time/trampoline.rs

```diff
@@ -1,35 +1,110 @@
-use crate::extn::core::time::Time;
+use crate::extn::core::time::{subsec::Subsec, Offset, Time};
 use crate::extn::prelude::*;
 
 // Constructor
-
 pub fn now(interp: &mut Artichoke) -> Result<Value, Error> {
-    let now = Time::now();
+    let now = Time::now()?;
     let result = Time::alloc_value(now, interp)?;
     Ok(result)
 }
 
-pub fn at(interp: &mut Artichoke, seconds: Value, microseconds: Option<Value>) -> Result<Value, Error> {
-    let seconds = implicitly_convert_to_int(interp, seconds)?;
-    let sub_second_nanos = if let Some(microseconds) = microseconds {
-        implicitly_convert_to_int(interp, microseconds)?
-            .checked_mul(i64::from(MICROS_IN_NANO))
-            .ok_or_else(|| ArgumentError::with_message("Time too large"))?
-    } else {
-        0_i64
-    };
+pub fn at(
+    interp: &mut Artichoke,
+    seconds: Value,
+    first: Option<Value>,
+    second: Option<Value>,
+    third: Option<Value>,
+) -> Result<Value, Error> {
+    // Coerce the params to the correct place. Specifically:
+    // - the options hash might not always be provided as the last argument.
+    // - subseconds can be provided with an optional symbol for the type of subsec.
+    //
+    // ```console
+    // [3.1.2] > Time.at(1)
+    // => 1970-01-01 01:00:01 +0100
+    // [3.1.2] > Time.at(1, 1)
+    // => 1970-01-01 01:00:01.000001 +0100
+    // [3.1.2] > Time.at(1, 1, :nsec)
+    // => 1970-01-01 01:00:01.000000001 +0100
+    // [3.1.2] > Time.at(1, in: "A")
+    // => 1970-01-01 01:00:01 +0100
+    // [3.1.2] > Time.at(1, 1, in: "A")
+    // => 1970-01-01 01:00:01.000001 +0100
+    // [3.1.2] > Time.at(1, 1, :nsec)
+    // => 1970-01-01 01:00:01.000000001 +0100
+    // [3.1.2] > Time.at(1, 1, :nsec, in: "A")
+    // => 1970-01-01 01:00:01.000000001 +0100
+    // ```
 
-    if let Some(time) = Time::at(seconds, sub_second_nanos) {
-        let result = Time::alloc_value(time, interp)?;
-        Ok(result)
+    let mut subsec = first;
+    let mut subsec_unit = second;
+    let mut options = third;
+
+    // Re-position the options hash under the `options` if it exists. Calling
+    // `Time.at` without the optional parameters will end up placing the
+    // options hash in the incorrect parameter position.
+    //
+    // ```console
+    // Time.at(0, in: "A")
+    // #          ^--first
+    // Time.at(0, 1, in: "A")
+    // #             ^-- second
+    // Time.at(0, 1, :nsec, in: "A")
+    // #                    ^-- third
+    // ```
+    //
+    // The below logic:
+    // - ensures the third parameter is a Ruby::Hash if provided.
+    // - if third param is not options, check the second paramter, if it is a
+    //   Ruby::Hash then assume this is the options hash, and clear out the
+    //   second parameter.
+    // - if second param is not options, check the first param, if it is a
+    //   Ruby::Hash then assume this is the options hash, and clear out the
+    //   first parameter.
```

Comment:
It appears that MRI supports this kind of syntax is slightly easier to parse way than mruby. I'm not sure where I spotted it, but it MRI had something around an `options hash`, where as ruby parses the named parameters as the last provided parameter instead. :shrug:.

I have a feeling the below might look a lot nicer with a `match`, but i'm open to ideas!

Reply:
yea Ruby 3 changed how kwargs are handled. They are no longer converted to `Hash`.

This is a good description of the changes: https://eregon.me/blog/2019/11/10/the-delegation-challenge-of-ruby27.html.